pkg/delete/delete.go

```diff
@@ -38,91 +35,79 @@ import (
 	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/delete/pointer"
 )
 
-type ClientSet struct {
-	Classic    client.ConfigClient
-	Settings   client.SettingsClient
-	Automation client.AutomationClient
-	Buckets    client.BucketClient
-	Documents  client.DocumentClient
-}
-
 type configurationType = string
 
 // DeleteEntries is a map of configuration type to slice of delete pointers
 type DeleteEntries = map[configurationType][]pointer.DeletePointer
 
 // Configs removes all given entriesToDelete from the Dynatrace environment the given client connects to
-func Configs(ctx context.Context, clients ClientSet, _ api.APIs, automationResources map[string]config.AutomationResource, entriesToDelete DeleteEntries) error {
-	copiedDeleteEntries := make(DeleteEntries)
-	for k, v := range entriesToDelete {
-		copiedDeleteEntries[k] = v
+func Configs(ctx context.Context, clients client.ClientSet, entriesToDelete DeleteEntries) error {
+	remainingEntriesToDelete, errCount := deleteAutomationConfigs(ctx, clients.AutClient, entriesToDelete)
+
+	//  Dashboard share settings cannot be deleted
+	if _, ok := remainingEntriesToDelete[api.DashboardShareSettings]; ok {
+		log.Warn("Classic config of type %s cannot be deleted. Note, that they can be removed by deleting the associated dashboard.", api.DashboardShareSettings)
+		delete(remainingEntriesToDelete, api.DashboardShareSettings)
+
+	}
+
+	// Delete rest of config types
+	for t, entries := range remainingEntriesToDelete {
+		if err := deleteConfig(ctx, clients, t, entries); err != nil {
+			log.WithFields(field.Error(err)).Error("Error during deletion: %v", err)
+			errCount += 1
+		}
 	}
 
-	var deleteErrors int
+	if errCount > 0 {
+		return fmt.Errorf("encountered %d errors", errCount)
+	}
+	return nil
+}
 
-	// Delete automation resources (in the specified order)
+func deleteAutomationConfigs(ctx context.Context, autClient client.AutomationClient, allEntries DeleteEntries) (DeleteEntries, int) {
+	remainingDeleteEntries := maps.Clone(allEntries)
+	errCount := 0
 	automationTypeOrder := []config.AutomationResource{config.Workflow, config.SchedulingRule, config.BusinessCalendar}
 	for _, key := range automationTypeOrder {
-		entries := copiedDeleteEntries[string(key)]
-		if clients.Automation == (*coreAutomation.Client)(nil) {
+		entries := allEntries[string(key)]
+		delete(remainingDeleteEntries, string(key))
+		if autClient == nil {
 			log.WithCtxFields(ctx).WithFields(field.Type(key)).Warn("Skipped deletion of %d Automation configuration(s) of type %q as API client was unavailable.", len(entries), key)
-			delete(copiedDeleteEntries, string(key))
 			continue
```

Comment:
I think this should be at least a return. Or we check for the null client in the Configs method.

Reply:
I had a closer look at the code, as I agreed with this comment. However, I think it's fine like this (for now). The reason is, that we report each config & component that it can't be deployed. A refactoring here that gives us the same log message information would be way more work and we should not do that right now.

Reply:
Let's revisit this when we fix logs and hopefully simplify the whole deletion code.